### ISSUE TITLE
Open new window in same directory

### ISSUE THREAD

**Issue by SevereOverfl0w** [jwilm/alacritty#808]
In other terminals, this is commonly bound to Ctrl-Shift-T.

Possibly a duplicate of https://github.com/jwilm/alacritty/issues/607 but I'm not certain.

**Comment by typekpb** [comment]
just a side note: <kbd>ctrl</kbd> + <kbd>shift</kbd> + <kbd>T</kbd> opens new tab in gnome-terminal for me (rather than new window)

kbd>ctrl</kbd> + <kbd>shift</kbd> + <kbd>n</kbd> does a new window for me

**Comment by jwilm** [comment]
#607 is different; it's focused on having multiple windows within the same process rather than being able to open additional windows.

Assuming you're on Linux/X11, there's a bit of a workaround you can do today since this isn't directly supported. I've got a script `spawn-alacritty-cwd` which achieves what you're looking for.

```sh
#!/bin/bash
# Spawn a new instance of Alacritty using the CWD of the currently focused
# Alacritty process.
#
# This is useful in environment like i3 where terminals are opened using a
# key-combination while another terminal is already focused.
#
# If the script is run with a non-Alacritty window in focus or a non-compliant
# version of Alacritty, an instance will be spawned in the user's $HOME.

ACTIVE_WINDOW=$(xdotool getactivewindow)
ACTIVE_WM_CLASS=$(xprop -id $ACTIVE_WINDOW | grep WM_CLASS)
if [[ $ACTIVE_WM_CLASS == *"Alacritty"* ]]
then
    # Get PID. If _NET_WM_PID isn't set, bail.
    PID=$(xprop -id $ACTIVE_WINDOW | grep _NET_WM_PID | grep -oP "\d+")
    if [[ "$PID" == "" ]]
    then
        exit 1
    fi
    # Get first child of terminal
    CHILD_PID=$(pgrep -P $PID)
    if [[ "$PID" == "" ]]
    then
        exit 1
    fi
    # Get current directory of child. The first child should be the shell.
    pushd "/proc/${CHILD_PID}/cwd"
    SHELL_CWD=$(pwd -P)
    popd
    # Start alacritty with the working directory
    alacritty --working-directory $SHELL_CWD
else
    alacritty
fi
```

Here's an example keybinding that uses the script:

```yaml
- { key: N, mods: Control|Shift, command: { program: "/bin/bash", args: ["-c", "/usr/local/bin/spawn-alacritty-cwd"] }
```

**Comment by SevereOverfl0w** [comment]
This looks really useful, I'll give it a go, thanks!

I think I meant Ctrl-Shift-N, yeah. These things end up burned in keymap memory.

**Comment by jwilm** [comment]
Er, I kind of hacked this up a bit. You may want to replace the other `exit 1` calls with `alacritty` calls.

**Comment by SevereOverfl0w** [comment]
For others following me:
1. there's a missing `}` in the above keybinding
2. I didn't need to use /bin/bash -c path-to-script, just /path-toscript
3. Using `xdo` I was able to shorten this to: `alacritty --working-directory "$(readlink -e /proc/"$(pgrep -P "$(xdo pid)" | head -n 1)"/cwd)"` only, I'm probably missing necessary error handling though :)

**Comment by jwilm** [comment]
The shortened version with `pgrep -P` and `xdo pid` make sense for the binding; nice work! The provided script will work even when not run as a child process, but if that's not necessary, your version is a lot nicer.

**Comment by SevereOverfl0w** [comment]
`alacritty --working-directory "$(readlink -e /proc/"$(pgrep -P "$(xdo pid)" | tail -n 1)"/cwd)"` 

For whatever reason, `tail` works better than `head` when you have a sub-sub-process. Previously the script failed if I had vim open (for some reason…).

This might be coincidental, and I may need to figure out if I actually need to filter the list of children somehow.

**Comment by SevereOverfl0w** [comment]
I, of course, immediately found a new edge-case under which it doesn't work, for which I cannot explain.

I've now got `alacritty --working-directory "$(readlink -e /proc/"$(pgrep -oP "$(xdo pid)")"/cwd)"` which has worked for my currently open terminals.

The `-o` finds the oldest pid with a parent of the current terminal. This seems to reliably be the shell PID.

I suspect it will break down if I open a sub-shell at some point, but I won't do that, not today.

**Comment by nixpulvis** [comment]
Seems to me a well thought out solution to #607 should indeed allow this to be fixed.

**Comment by rubdos** [comment]
Is there a workaround/hack for Wayland, like the script above? I'm on `sway`, fyi.

**Comment by neon64** [comment]
@rubdos the output of `swaymsg -t get_tree` seems to have all the information you need in order to implement it on Wayland, except just parsing the nested JSON structure could be a pain and/or quite inefficient. I'll try and botch together an example now.

**Comment by neon64** [comment]
Okay so I came up with this which is basically just jwilm's script but written in rust and using the `i3ipc` crate to get information about which process is currently focused. Unfortunately I needed to modify the `i3ipc` crate to include the 'pid' field that I wanted to use, so perhaps its an undocumented part of Sway's IPC.

https://github.com/neon64/alacritty_new_window_cwd

**Comment by rubdos** [comment]
That's indeed always been the issue when I tried implementing something like that: getting the PID field. This indeed works, but let's hope that this issue gets resolved nonetheless :-)

**Comment by jwilm** [comment]
It sounds like the ask here is that we might want to add an Action which spawns a new alacritty process using the child process' CWD? Then we can get rid of WM-specific tools like the script I use and the progrgam neon64 provided.

**Comment by chrisduerr** [comment]
Yeah, rather than putting a lot of effort in tools working around this, I'd just recommend to look into adding an additional action. I'm happy to help here if there are any questions, in theory this shouldn't be very difficult.

**Comment by Camotubi** [comment]
Hi guys, I think i got this to work. I added a new action to open a new instance of alacritty with the cwd. I dont have macos or Windows machine right now so it wont work probably. I would like you to review my changes. If it is ok, I could make a pull request.
To test the action I added this line to my config file:
 { key: N,        mods: Alt,              action: NewInstanceSameDir  }
Here are my changes:
https://github.com/Camotubi/alacritty/commit/4be24763a90905dbdab9bdcb4e66e72d3103476c

**Comment by Camotubi** [comment]
> @Camotubi If you've got a working version feel free to just send a PR. Feedback and improvement can then be handled there separately without losing track of things! :)

Done :D

**Comment by dm1try** [comment]
btw, `osx` users might be interested in the AppleScript solution 😎
```applescript
-- /Users/iam/scripts/new_alacritty_with_cwd.scpt

tell application "System Events"
	-- get PID of the current active Alacritty window
	set activeAlacrittyPID to get unix id of first application process whose name is "Alacritty" and frontmost = true
	-- get PID of the current active shell
	set shellPID to do shell script "pgrep -P " & activeAlacrittyPID & " -a zsh"
	-- get the current working directory
	set shellCWD to do shell script "lsof -a -d cwd -p " & shellPID & " -F n | cut -c 2- | tail -n1"
	-- run a new Alacritty instance within the current working directory
	do shell script "open -nb io.alacritty --args --working-directory " & shellCWD
end tell
```
```yaml
# ~/.alacritty.yml
  - { key: N,        mods: Command, command: {program: "osascript", args: ["/Users/iam/scripts/new_alacritty_with_cwd.scpt"]} }
```

**Comment by wucke13** [comment]
Also, have a look at https://github.com/schischi/xcwd . It lets you start a terminal in the CWD of the currently focused X window (not necessarly `alacritty`).

**Comment by teodorlu** [comment]
@wucke13 Thanks a lot! You implicitly solved my real program -- not just for the terminal. Thanks!

**Comment by fedoranvar** [comment]
also solution, is to set bindkey for function:

```
# define function for creating terminal
newterm_curr_cd() {
    nohup alacritty --working-directory $(pwd) &
}

bindkey -s '^t' 'newterm_curr_cd\n'  # CTRL-T in terminal calls for newterm_curr_cd function

```

**Comment by Rivers47** [comment]
The default installation config on the distro I use doesn't have the keybinding commented, in case anyone is wondering, you just have to add { key: N,        mods: Control|Shift,    action: SpawnNewInstance    } to the keybindings section in your config.yaml, the ctrl+shift+n spawns a new window in the same working directory.

**Comment by muralikodali** [comment]
I am on Arch Linux and my alacritty version is  0.6.0 (04cbf767).

I have added the keybinding { key: N, mods: Control|Shift, action: SpawnNewInstance }. It is not spawning any new window in the same directory. 

I have also tried xcwd without success in alacritty,  which is working well with termite and urxvt.

**Comment by wucke13** [comment]
@muralikodali You missed to specify one very important piece of information here. Does no window spawn, or does the window spawn in the wrong directory? How did you setup `xcwd`, for it to fail? I'm sorry to hear that things don't work for you _yet_, but you got to put in a little more information to receive help.

**Comment by teodorlu** [comment]
@muralikodali 

I'm running

    alacritty --working-directory `xcwd`

, and new terminals open in the same folder. You'll probably have to describe what you're trying to get any good feedback.

**Comment by SLAYER-CODE** [comment]
alacritty --working-directory $("pwd")

**Comment by JacobTravers** [comment]
On Fish I tried the method above on my own before coming here, and I had added `&` to run in the background:
```fish
alacritty --working-directory (pwd) &
```
However this spawns a new terminal as a child-process of the current one, so closing any parent-terminal will result in the removal of all its children, it is also not easy to incorporate as a keybinding on spectrwm (my wm).
 
To get around this I just created a `new-instance.sh` and then added a binding to the script in my `spectrwm.conf`
```sh
#!/bin/sh
alacritty --working-directory `xcwd`
```

**Comment by g1sbi** [comment]
@JacobTravers I tried your solution, and while alacritty opens correctly a new window in the correct directory, I get an error saying "Invalid working directory: "xcwd"

**Comment by JacobTravers** [comment]
@g1sbi it sounds like you might be overlapping a keybinding or something... It should only have one of those behaviors and not both simultaneously.

Also I've updated my comment since I learned more about scripting. Please read it again and try the updated scripts. If you are still having issues:

Make sure `xcwd` is in your `$PATH`. Check with `which xcwd`.

Then `cd` into any directory and run this:
```fish
sh -c 'alacritty --working-directory "$(xcwd)"'
```
if it works, you are likely doing something wrong elsewhere and it is not the script.

**Comment by stefanv** [comment]
With sway, you can use [this implementation of xcwd](https://www.reddit.com/r/swaywm/comments/ayedi1/opening_terminals_at_the_same_directory/ei7i1dl/), and add this to your config:

```
set $term alacritty --working-directory "$(sway-cwd)"
bindsym $mod+t exec $term
```

**Comment by etherrorcode404** [comment]
With Hyprland, you can use [this implementation of xcwd](https://github.com/vilari-mickopf/hyprcwd), with the following hypr config:
 
 ```
$term = alacritty --working-directory "$(hyprcwd)"
bind = $mod, RETURN, exec, $term
 ```

**Comment by judecabra** [comment]
that's my way.

in alacritty.yml:

key_bindings:
  - { key: Return, mods: Control, action: SpawnNewInstance }
commands:
  - { name: SpawnNewInstance, command: alacritty & disown }

**Comment by fercomunello** [comment]
I simply create a dot alias for that, yes just a dot! =) 
https://github.com/fercomunello/dotfiles/commit/bda05732836a1ca677712d292c76aa2998debe05

```sh
# Open the terminal on the same directory in other window / unix process
alias .='alacritty --working-directory=$(pwd) & disown'
```

**Comment by retsi101** [comment]
> #607 is different; it's focused on having multiple windows within the same process rather than being able to open additional windows.
> 
> Assuming you're on Linux/X11, there's a bit of a workaround you can do today since this isn't directly supported. I've got a script `spawn-alacritty-cwd` which achieves what you're looking for.
> 
> ```shell
> #!/bin/bash
> # Spawn a new instance of Alacritty using the CWD of the currently focused
> # Alacritty process.
> #
> # This is useful in environment like i3 where terminals are opened using a
> # key-combination while another terminal is already focused.
> #
> # If the script is run with a non-Alacritty window in focus or a non-compliant
> # version of Alacritty, an instance will be spawned in the user's $HOME.
> 
> ACTIVE_WINDOW=$(xdotool getactivewindow)
> ACTIVE_WM_CLASS=$(xprop -id $ACTIVE_WINDOW | grep WM_CLASS)
> if [[ $ACTIVE_WM_CLASS == *"Alacritty"* ]]
> then
>     # Get PID. If _NET_WM_PID isn't set, bail.
>     PID=$(xprop -id $ACTIVE_WINDOW | grep _NET_WM_PID | grep -oP "\d+")
>     if [[ "$PID" == "" ]]
>     then
>         alacritty
>     fi
>     # Get first child of terminal
>     CHILD_PID=$(pgrep -P $PID)
>     if [[ "$PID" == "" ]]
>     then
>         alacritty
>     fi
>     # Get current directory of child. The first child should be the shell.
>     pushd "/proc/${CHILD_PID}/cwd"
>     SHELL_CWD=$(pwd -P)
>     popd
>     # Start alacritty with the working directory
>     alacritty --working-directory $SHELL_CWD
> else
>     alacritty
> fi
> ```
> 
> Here's an example keybinding that uses the script:
> 
> ```yaml
> - { key: N, mods: Control|Shift, command: { program: "/bin/bash", args: ["-c", "/usr/local/bin/spawn-alacritty-cwd"] }
> ```

I'm trying something like this on Windows 10

[[keyboard.bindings]]
  key = "N"
  mods = "Alt"
  command = { program = "C:/Program Files/PowerShell/7/pwsh.exe", args = ["--no-exit", "--command", "cd ."] }
  
But nothing happens, no error messages are shown.

Am I missing something?

TIA

**Comment by yki-t** [comment]
Has anyone been able to achieve this with Windows WSL?
This feature worked on Linux and Mac with some tricks, but not on Windows when setting WSL as the default shell with this config

```toml
[shell]
program = "wsl"
```

The approach to infer the CWD from the PID of the active window (solution for Mac) didn't success. This is because Windows likely doesn't know the Shell PID of WSL (of alacritty of Windows).

**Comment by nendonerd** [comment]
For MacOS users, I found a way to press a shortcut combo to open a new Alacritty window with the cwd of current Alacritty window, and also focus the new window:
1. add a zsh hook to update the title of the Alacritty window to the latest cwd
```zsh
##########
# .zshrc #
##########

# Define the function to update the terminal title
function update_title {
  print -Pn "\e]2;$PWD\a"
}

# Add the function to the chpwd hook array
autoload -U add-zsh-hook
add-zsh-hook chpwd update_title

# Call the function to set the initial title
update_title

```
2. bind hyper-enter to open Alacritty with the current directory of the active Alacritty window and focus it
```zsh
##########
# skhdrc #
##########

hyper - return : ~/.config/skhd/openAlacritty.sh
```
```zsh
####################
# openAlacritty.sh #
####################

# get yabai id of the current active Alacritty window
alacritty_window=$(yabai -m query --windows | jq -r '.[] | select(.app == "Alacritty" and .["has-focus"] == true) | .id')

# if such id exist, get the title of that window as cwd
if [ -n "$alacritty_window" ]; then
  current_dir=$(yabai -m query --windows --window $alacritty_window | jq -r '.title')
  open -na "Alacritty" --args --working-directory "$current_dir"
else
  open -na "Alacritty"
fi

# focus on the new window
sleep 0.5 # wait for the window to open
new_window_id=$(yabai -m query --windows | jq -r '.[] | select(.app == "Alacritty" and .["has-focus"] == true) | .id' | tail -n 1)
yabai -m window --focus "$new_window_id"
```

**Comment by crossz** [comment]
On windows, using zsh, 
just add an alias in `~/.zshrc`

```
alias aw="alacritty --working-directory ."
```

then, afterwards, command `aw` will start a new alacritty window with same path.

**Comment by stefanv** [comment]
> then, afterwards, command `aw` will start a new alacritty window with same path.

From within the terminal, that works. But the question here was how to have a command that would work with a key binding.

**Comment by kchibisov** [comment]
Default bindings account for the working directory out of the box. Unless you use `tmux` or some other multiplexer.

**Comment by olivia-fl** [comment]
A problem with the approaches that rely on using the focused window PID from the WM is that if you are using alacritty in daemon mode all windows will be the same alacritty process.

An IPC message like `alacritty msg get-focused-pid` that returns the PID of the focused window's child process might help with this. I think ideally I'd want something like `get-window-pid` that takes an platform-specific identifier for a window. I'm not sure that's implementable on wayland though, and don't know anything about osx/windows.

**Comment by khollbach** [comment]
For niri specifically, here's a script to open new Alacritty windows in the same working directory as the previous window: [niri-alacritty-recent-cwd](https://github.com/khollbach/niri-alacritty-recent-cwd)
